### PR TITLE
[rllib] Q-Mix implementation (Q-Mix, VDN, IQN, and Ape-X variants)

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Adapted from https://github.com/oxwhirl/pymarl_alpha/tree/master/src

TODO:
- [x] Add env grouping wrapper
- [x] Use tuple action spaces / handle obs batching properly
- [x] Handle action masking
- [x] Validate it reaches reward of 8 in the two-step game
- [x] Enable batch replay in Ape-X
- [x] Docs

Sanity checked using the two-step game example from paper. Overall Q-loss is lower with a mixer enabled (though it seems like we converged to 8 rewards in all cases anyways as epsilon anneals to 0).
qmix

It's likely there are some issues still, but those will be smoked out with the Starcraft examples.

## Related issue number

https://github.com/ray-project/ray/issues/3540
Closes https://github.com/ray-project/ray/pull/3547
Closes https://github.com/ray-project/ray/issues/3546

### CODE REVIEW

File: python/ray/rllib/env/async_vector_env.py [hunk 1]
```diff
@@ -304,9 +304,9 @@ def send_actions(self, action_dict):
                 raise ValueError(
                     "Key set for obs and rewards must be the same: "
                     "{} vs {}".format(obs.keys(), rewards.keys()))
-            if set(obs.keys()) != set(infos.keys()):
-                raise ValueError("Key set for obs and infos must be the same: "
-                                 "{} vs {}".format(obs.keys(), infos.keys()))
+            if set(infos.keys()).difference(set(obs.keys())):
```
Comment (richardliaw) [inline python/ray/rllib/env/async_vector_env.py]: I think you can do `set(info).difference(set(obs))`

File: doc/source/rllib-algorithms.rst [hunk 2]
```diff
@@ -245,3 +245,18 @@ Tuned examples: `Humanoid-v1 <https://github.com/ray-project/ray/blob/master/pyt
    :language: python
    :start-after: __sphinx_doc_begin__
    :end-before: __sphinx_doc_end__
+
+QMIX Monotonic Value Factorisation (QMIX, VDN, IQN)
+---------------------------------------------------
+`[paper] <https://arxiv.org/abs/1803.11485>`__ `[implementation] <https://github.com/ray-project/ray/blob/master/python/ray/rllib/agents/qmix/qmix.py>`__ Q-Mix is a specialized multi-agent algorithm. Code here is adapted from https://github.com/oxwhirl/pymarl_alpha  to integrate with RLlib multi-agent APIs. To use Q-Mix, you must specify an agent `grouping <https://github.com/ray-project/ray/blob/master/python/ray/rllib/env/group_agents_wrapper.py>`__ in the environment (see the `two-step game example <https://github.com/ray-project/ray/blob/master/python/ray/rllib/examples/twostep_game.py>`__). Currently, all agents in the group must be homogeneous. The algorithm can be scaled by increasing the number of workers or using Ape-X.
+
+Q-Mix is implemented in `PyTorch <https://github.com/ray-project/ray/blob/master/python/ray/rllib/agents/qmix/qmix_policy_graph.py>`__ and is currently *experimental*.
```
Comment (richardliaw) [inline doc/source/rllib-algorithms.rst]: Maybe tag the experimental comment with `.. note:` or `.. warning` so that it is more visible.

File: python/ray/rllib/env/constants.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+# info key for the individual infos of an agent, for example:
+# info: {
+#   group_1: {
+#      _group_infos: [{"_avail_actions": ...}, {}],  # 2 agents in this group
+#   }
+# }
+GROUP_INFO = "_group_info"
+
+# info key for the available actions mask, for example:
```
Comment (richardliaw) [inline python/ray/rllib/env/constants.py]: Is this the intended way to specify the action mask? Doesn't the documentation recommend it going into the observation?

```
The environment should return a mask and/or list of valid action embeddings as part of the observation for each step. To enable batching, the number of actions can be allowed to vary from 1 to some max number:
```

File: python/ray/rllib/agents/qmix/mixers.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import torch as th
```
Comment (richardliaw) [inline python/ray/rllib/agents/qmix/mixers.py]: __future__?

File: python/ray/rllib/agents/qmix/model.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from torch import nn
+import torch.nn.functional as F
+
+
+# TODO(ekl) we should have common models for pytorch like we do for TF
```
Comment (richardliaw) [inline python/ray/rllib/agents/qmix/model.py]: Is that different from `rllib/models/pytorch/`?

File: python/ray/rllib/optimizers/async_replay_optimizer.py [hunk 6]
```diff
@@ -334,8 +394,8 @@ def step(self):
                 grad_out = self.local_evaluator.compute_apply(replay)
                 for pid, info in grad_out.items():
                     prio_dict[pid] = (
-                        replay.policy_batches[pid]["batch_indexes"],
-                        info["td_error"])
+                        replay.policy_batches[pid].data.get("batch_indexes"),
+                        info.get("td_error"))
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/async_replay_optimizer.py]: do you need to set defaults or will None be fine

File: python/ray/rllib/optimizers/async_replay_optimizer.py [hunk 7]
```diff
@@ -299,6 +311,54 @@ def stats(self, debug=False):
...
 
 
+@ray.remote(num_cpus=0)
+class BatchReplayActor(object):
+    """The batch replay version of the replay actor.
+
+    This allows for RNN models, but ignores prioritization params.
+    """
+
+    def __init__(self, num_shards, learning_starts, buffer_size,
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/async_replay_optimizer.py]: can you document?

File: python/ray/rllib/optimizers/sync_replay_optimizer.py [hunk 8]
```diff
@@ -159,13 +157,13 @@ def _replay(self):
...
-            samples[policy_id] = SampleBatch({
-                "obs": obses_t,
-                "actions": actions,
-                "rewards": rewards,
-                "new_obs": obses_tp1,
-                "dones": dones,
-                "weights": weights,
-                "batch_indexes": batch_indexes
-            })
+                samples[policy_id] = SampleBatch({
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/sync_replay_optimizer.py]: was this a bug before?

File: python/ray/rllib/evaluation/policy_graph.py [hunk 9]
```diff
@@ -42,17 +42,21 @@ def compute_actions(self,
...
+                        episodes=None,
+                        **kwargs):
         """Compute actions for the current policy.
 
         Arguments:
             obs_batch (np.ndarray): batch of observations
             state_batches (list): list of RNN state input batches, if any
             prev_action_batch (np.ndarray): batch of previous action values
             prev_reward_batch (np.ndarray): batch of previous rewards
+            info_batch (np.array): batch of info objects
```
Comment (richardliaw) [inline python/ray/rllib/evaluation/policy_graph.py]: this is just a list right?

File: python/ray/rllib/env/group_agents_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+from __future__ import print_function
+
+from collections import OrderedDict
+
+from ray.rllib.env.constants import GROUP_REWARDS, GROUP_INFO
+from ray.rllib.env.multi_agent_env import MultiAgentEnv
+
+
+class _GroupAgentsWrapper(MultiAgentEnv):
+    """Wraps a MultiAgentEnv environment with agents grouped as specified.
```
Comment (richardliaw) [inline python/ray/rllib/env/group_agents_wrapper.py]: maybe add a TODO for unit testing this

File: python/ray/rllib/evaluation/policy_graph.py (hunk 9 above)
Comment (ericl) [reply python/ray/rllib/evaluation/policy_graph.py]: yeah "batch" = list

File: python/ray/rllib/optimizers/sync_replay_optimizer.py (hunk 8 above)
Comment (ericl) [reply python/ray/rllib/optimizers/sync_replay_optimizer.py]: yeah

File: python/ray/rllib/env/group_agents_wrapper.py (hunk 10 above)
Comment (ericl) [reply python/ray/rllib/env/group_agents_wrapper.py]: done

**Review by richardliaw — APPROVED** [review]
looks good from high level, will integrate into starcraft example

File: doc/source/rllib-algorithms.rst (hunk 2 above)
Comment (ericl) [reply doc/source/rllib-algorithms.rst]: It's probably fine, it won't be experimental for long.

File: python/ray/rllib/agents/qmix/model.py (hunk 5 above)
Comment (ericl) [reply python/ray/rllib/agents/qmix/model.py]: Yeah but that's pretty old code. It will probably have to be cleaned up a bunch to get it working properly with the newer APIs.

File: python/ray/rllib/optimizers/async_replay_optimizer.py (hunk 6 above)
Comment (ericl) [reply python/ray/rllib/optimizers/async_replay_optimizer.py]: default is None

File: python/ray/rllib/optimizers/async_replay_optimizer.py (hunk 7 above)
Comment (ericl) [reply python/ray/rllib/optimizers/async_replay_optimizer.py]: No, it's not worth it for the internal optimizers. They are documented in the agent apis.

**Comment by ericl** [comment]
Ok, changed to move `avail_actions` to a dict observation.

File: python/ray/rllib/evaluation/policy_graph.py (hunk 9 above)
Comment (richardliaw) [reply python/ray/rllib/evaluation/policy_graph.py]: `(np.array)` -> `list`
